serverside-javascript-best-practices.md

```diff
@@ -320,6 +350,53 @@ createBooking.call(self, product, params, 'upgrades', 'supplier', 'payment', fun
 
 
 
+## Handling heavily sequential async code with promises
+
+* !! Using promises for heavily sequential async code has the downside of hiding the data being passed between functions. We lose the ability to wrap each layer in a closure to selectively pass the correct data betweeen the correct entites. Eveything must flow through the middle to get to the end. This breaks our rules on single-responsibility, but allows for slightly cleaner orchestration:
+* !! Sinon promises do not support sandboxing yet, so you'll have to manually restore everything by hand
+* !! Its impossible to prove that the data flow between promises is correct, only that data moves between them.
+
+```javascript
+createBooking._orchestrate = function(data) {
+  return createBooking._sendBook1Request(data)
+    .then(createBooking._generatePaymentRequest)
+    .then(createBooking._processPayment)
+    .then(createBooking._sendBook2Request)
+    .then(createBooking._generateBookingReference);
+};
+```
+```javascript
+require('sinon-bluebird');
+
+sinon.stub(createBooking, '_sendBook1Request').resolves('data1');
+sinon.stub(createBooking, '_generatePaymentRequest').resolves('data2');
+sinon.stub(createBooking, '_processPayment').resolves('data3');
+sinon.stub(createBooking, '_sendBook2Request').resolves('data4');
+sinon.stub(createBooking, '_generateBookingReference').resolves('data5');
+
```

Comment:
Looks tidy in a basic use case but once you start modelling a more real world example things will get messy pretty quickly.  Stubbing objects and multiple arguments for more than one function at a time is gonna get out of hand, not sure that's avoidable though.  I'd suggest using `.spread` instead of `.then` for this kind of pattern, gives greater visibility on what's being passed between the methods.